Transport-abstracted signature policy

mtrmac opened the pull request:
@runcom PTAL . This is by no means finished, but shows the direction I’m planning to use, early feedback on the API would help me avoid a dead end.  Start with `types.go` I think, I will comment more on the code.

mtrmac commented:
See also https://github.com/projectatomic/skopeo/pull/138 for the consumer.

mtrmac commented on transports.go:
```diff
@@ -0,0 +1,30 @@
+		docker.Transport,
+		oci.Transport,
+		openshift.Transport,
+	} {
```
This would nicely fit into `image`, but it needs to be a separate package because these transport-dependent packages depend on `image`, which would cause a circular dependency.

Adding a `containers/image/transports` for this small file feels a bit silly, and having the top-level `containers/image` package point at the contained transports seems to make sense — OTOH using both `containers/image` and `containers/image/image` in a single source file then ends up pretty very ugly.

mtrmac commented on transports_test.go:
```diff
@@ -0,0 +1,12 @@
+
+func TestKnownTransports(t *testing.T) {
+	assert.NotNil(t, KnownTransports) // Ensure that the initialization has actually been run
+	assert.True(t, len(KnownTransports) > 1)
```
(Surprisingly enough, I have actually managed to write the initialization incorrectly the first time. Hence this test.)

mtrmac commented on types/types.go:
```diff
@@ -5,6 +5,46 @@ import (
 
+// ImageTransport is a top-level namespace for ways to refer to an image.
+// It should generally correspond to ImageSource/ImageDestination implementations.
+//
```
I.e. ImageTransport is our `dir:`/`docker:` etc.

mtrmac commented on types/types.go:
```diff
@@ -5,6 +5,46 @@ import (
+	Name() string
+	// ParseReference converts a string, which should not start with the ImageTransport.Name prefix, into an ImageReference.
+	ParseReference(reference string) (ImageReference, error)
+}
```
Planned extension: ```ValidateNamespaceString```, validates that the input is either a valid reference or a valid namespace containing references, to be used in something like `policySpecificMap.UnmarshalJSON`.

mtrmac commented on types/types.go:
```diff
@@ -5,6 +5,46 @@ import (
+}
+
+// ImageReference is an abstracted way to refer to an image or a namespace containing images, namespaced within an ImageTransport
+type ImageReference interface {
```
The signature code would stop talking about Docker References, and start using this.

mtrmac replied:
(or mostly stop; I don’t quite think that creating a signature claiming an image is `dir:/home/mitr/image` is at all useful.)

runcom replied:
well, we already stated somewhere is by any mean to be used other than a simple debug functionality so I'm fine with this. We probably need to make `dir:` hidden under `--debug` flag so it can only be used if running in debug (no need to think about this now)

mtrmac commented on types/types.go:
```diff
@@ -5,6 +5,46 @@ import (
+	// This is not purely for display: parsing the string into an ImageReference must result in an equivalent reference.
+	// It is highly recommended for FullyExplicitString to be designed so that all parent namespaces of a reference are initial prefixes of FullyExplicitString().
+	// It is perfectly acceptable for ConvenientString() and FullyExplicitString() to return the same value if the underlying references do not have format variants.
+	FullyExplicitString() (string, error)
```
The definition of these may need fine-tuning; they should allow replacing `IntendedDockerReference` and `CanonicalDockerReference`.

mtrmac replied:
`Image{,Source,Destination}` would then, instead of providing `IntendedDockerReference` and `CanonicalDockerReference`, simply provide a `Reference` returning what they were created with, and the mapping intelligence would all be in the `ImageReference` implementation.

runcom replied:
+1

mtrmac commented on types/types.go:
```diff
@@ -5,6 +5,46 @@ import (
+	NewImageSource(certPath string, tlsVerify bool) (ImageSource, error)
+	// NewImageDestination returns a types.ImageDestination for this reference.
+	NewImageDestination(certPath string, tlsVerify bool) (ImageDestination, error)
+}
```
Planned extension: `ParentNamespaces() []string`, returning a list of namespaces to possibly use in the policy.

mtrmac commented on docker/docker_image.go:
```diff
@@ -19,7 +19,17 @@ type Image struct {
+		return nil, err
+	}
+	return newDockerImage(ref, certPath, tlsVerify)
+}
```
This PR adds these `parseReference`+`new${x}Image${y}` helpers all over the place. I am not sure that this makes much sense. Yes, it is nice not to have to do any parsing in the caller… OTOH there are now so many similar and redundant ways to get from a string to an `types.Image$something`.

Arguably anyone starting with a string should go through https://github.com/projectatomic/skopeo/pull/138/files#diff-6179b09a11bd9337d53f297f715ac27dR13 (or an equivalent in this repo); and any caller knowledgeable enoug about a transport to directoy call one of these functions might be knowledgeable enough to give it a fully parsed specification (`reference.Named` + tag for Docker, `namespace`, `stream`, `tag` for OpenShift) etc.

I’ll take a look at this tomorrow with fresh eyes, comments very much apprecireated.

runcom replied:
I was feeling the same way (when working on client abstraction)

> and any caller knowledgeable enough about a transport to directly call one of these functions might be knowledgeable enough to give it a fully parsed specification (reference.Named + tag for Docker, namespace, stream, tag for OpenShift) etc.

+1

> Arguably anyone starting with a string should go through https://github.com/projectatomic/skopeo/pull/138/files#diff-6179b09a11bd9337d53f297f715ac27dR13

and +1

mtrmac commented on docker/docker_image_dest.go:
```diff
@@ -8,31 +8,33 @@ import (
-	ref reference.Named
-	tag string
-	c   *dockerClient
+	dockerReference
```
Using unnamed members like this now to minimize clutter in the patch, I guess this should use a real named member in the future for clarity.

runcom replied:
Agree

mtrmac commented on docker/docker_image_dest.go:
```diff
@@ -8,31 +8,33 @@ import (
 // NewDockerImageDestination creates a new ImageDestination for the specified image and connection specification.
 func NewDockerImageDestination(img, certPath string, tlsVerify bool) (types.ImageDestination, error) {
-	ref, tag, err := parseDockerImageName(img)
+	ref, err := parseReference("//" + img)
```
(I will pretty soon submit a PR to make these `//` optional everywhere.)

mtrmac commented on docker/docker_transport.go:
```diff
@@ -0,0 +1,120 @@
+// It is highly recommended for FullyExplicitString to be designed so that all parent namespaces of a reference are initial prefixes of FullyExplicitString().
+// It is perfectly acceptable for ConvenientString() and FullyExplicitString() to return the same value if the underlying references do not have format variants.
+func (ref dockerReference) FullyExplicitString() (string, error) {
+	// FIXME: Is this the right thing to do?
```
This originates, will replace, (and, via the interface, generalize) `policy_eval.go:fullyExpandedDockerReference`.

mtrmac commented on docker/docker_transport.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+// ParseReference converts a string, which should not start with the ImageTransport.Name prefix, into an dockerReference.
+func parseReference(refString string) (dockerReference, error) {
```
Three things called parseReference in a single file, _and_ a `NewReference` to boot. Feels a bit silly. Yes, it would make callers easier to read, not sure that it is worth it.

runcom replied:
I'm fine with this, we can reiterate as soon as we find a good ground for the exported functions

mtrmac commented on docker/docker_utils.go:
```diff
@@ -1,20 +0,0 @@
-
-// parseDockerImageName converts a string into a reference and tag value.
-func parseDockerImageName(img string) (reference.Named, string, error) {
-	ref, err := reference.ParseNamed(img)
```
Note how this disappears.

runcom commented on oci/oci_transport.go:
```diff
@@ -0,0 +1,107 @@
+	if sep == -1 {
+		dir = reference
+		tag = "latest"
+	} else {
```
Great I forgot to fix this the last time

runcom commented:
Generally, based on comments in the code also, seems like a very well welcome improvement and it makes up space for cleaning other interface parts (like `IntentedDockerReference` etc etc)

So I'm +1 here

mtrmac commented:
Updated:

- Merges/includes #39, so that it can be reviewed in pieces. I will rebase as necessary.
- Changed naming, now the methods are explicitly `PolicyConfigurationIdentity`/`PolicyConfigurationNamespaces`, naming things for what the intent and promised properties are, instead of vague/too general `FullyExplicitString`.
- Now includes the necessary policy configuration updates.
- ''Note'': `oci:` now prohibits `:` characters anywhere in the path, otherwise policy scopes like `/some/path:with` become ambiguous: is this an image at `/some/path` with tag `:with`, or a namespace for images in subdirectories of `/some/path:with`, using any tag? . I am not sure that this restriction reasonable, we could instead change the policy configuration naming for `oci`, e.g. to be `/full/path/:tag`, requiring a `/:` to separate path and tag.

WIP because I need to re-integrate into mtrmac/image:integrate-all-the-things before being 100% certain, a corresponding skopeo branch is missing, and it seems highly desirable to validate policy scope names when parsing the policy; but it is very unlikely that anything in this PR will change.

mtrmac commented:
Rebased and ready for individual review; I will add scope name verification when parsing the policy.

mtrmac commented:
Scope name validation added; works with
```
make test-skopeo SKOPEO_REPO=mtrmac/skopeo SKOPEO_BRANCH=location-namespaced-signatures
```

@runcom PTAL

runcom commented on oci/oci_transport.go:
```diff
@@ -47,12 +54,24 @@ func ParseReference(reference string) (types.ImageReference, error) {
+	// This is necessary to prevent directory paths returned by PolicyConfigurationNamespaces
+	// from being ambiguous with values of PolicyConfigurationIdentity.
+	if strings.Contains(resolved, ":") {
+		return nil, fmt.Errorf("Invalid OCI reference %s:%s: path %s contains a colon", dir, tag, resolved)
```
I think this ok now. Eventually this OCI isn't mean for _directory_ like output like `dir` - that means `oci_dest.go` will work exactly as `docker_img_dest.go` works (remote services). I have a TODO to create a new OCITar destination probably which will adhere to this and not allow colon in path.

mtrmac replied:
Does “This is OK now" mean “this naming restriction is acceptable” or “we need to allow such paths”?

runcom replied:
acceptable, I don't see why not. OCI_dest right now is kind of the same as dir - OCI image transportation and distribution isn't even yet defined so I treat oci: as dir: and this lgtm

runcom commented on private/dir.go:
```diff
@@ -0,0 +1,55 @@
+package private
```
As a non native english speaker, I find `private` to be too generic and I've honestly not fully understood the purpose of this package other than path sanitization. Could you either explain why this `private` name or add a package comment to explain?

runcom replied:
I'm not sure I explained myself well but `private.ResolvePathToFullyExplicit` doesn't tell me much about `private`

mtrmac replied:
There is `doc.go`.

`private` is simply a grab-bag of stuff which is useful to share between transports, primarily for policy configuration scoping (Docker + OpenShift share Docker Reference formatting, dir + OCI share path canonicalization).

runcom replied:
I just don't like `private` as a oo like mean of "do not use this" because it's not how Go works unfortunately :( as said below, I'd better have different packages with meaningful names (maybe under `transports` also if they're related to transports

runcom replied:
> There is doc.go.

sorry I'm going commit by commit and doc.go was in the next commit

mtrmac replied:
> I just don't like private as a oo like mean of "do not use this"

That’s fair; OTOH the semantics of what these utils do is very much targeted at `PolicyConfigurationIdentity` etc., very unlikely to be useful for anything else (well, `resolveExistingPathToFullyExplicit` could be useful, aka `canonicalize_file_name(3)`, but we don’t export _that_ one and it doesn’t belong into `containers/image` as a public API conceptually), and I have no intention of maintaining these functions as an API in any way useful to the general public for any other purpose than implementing `ImageTransport.PolicyConfiguration…`.  So the `private` subpackage name was intended to say “this is not for you, do not touch”; if language access controls can’t enforce it, we can still warn humans :)

runcom replied:
> if language access controls can’t enforce it, we can still warn humans :)

right, I was arguing against Go as well :) but still, either have `pkg/` or move them somewhere else w/o `private` looks better to me!

runcom commented on private/doc.go:
```diff
@@ -0,0 +1,3 @@
+// Package private contains helpers private to github.com/containers/image, with no intent or commitment to make them useful for any other purpose.
+// Do not use outside!
+package private
```
Ok I found this.

runcom replied:
so, I'd rather have a `pkg/MEANINGFULNAME` for each stuff in this package instead of a catch-all-oo-utils-like packge

mtrmac replied:
Sure, will do. `pkg/private/explicitfilepath` and `pkg/private/dockerpolicyconfiguration` ?  Or do you want it without the `private` component?

runcom replied:
w/o `private` component and probably under `transport` (IIUC they're related to transports, `explicitfilepath`) and/or `docker` (`dockerpolicyconfiguration` calling it just `policyconfiguration`)

mtrmac replied:
> and/or `docker`

Ooh, yes; we can treat this as “layered transport“ (OCI, OpenShift) reusing functionality from an underlying one (dir, Docker). That’s actually pretty clean, good idea.

runcom replied:
yep, well yeah kind of, not sure it's fully feasible cleanly :smile: but yeah

mtrmac replied:
Yeah, kind of; they would obviously not be actually inheriting from each, other, but “I am the OpenShift transport, I use a dockerImageSource for some operations, and I import docker/policyconfiguration for some others” is a readable story.

runcom replied:
yep it is, I like it

runcom commented:
The rest LGTM - if I understand correctly the last 2 commits leave some todos around also

runcom commented:
lgtm

mtrmac commented:
> The rest LGTM - if I understand correctly the last 2 commits leave some todos around also

If you mean https://github.com/containers/image/pull/33/commits/cd5e886c09225e1ea2e39a264304e4337dca9d98 , that `"docker"` lookup is replaced by https://github.com/containers/image/pull/33/commits/77f24aa23d20954cb433953bde9642f0e94522a9 .

https://github.com/containers/image/pull/33/commits/0ac31b213b6e2ff7806eb5f227bb9cf4b1238834 is left as a todo (or rather, `FIXME?`), yes.

mtrmac commented:
Updated:

- `private.ResolvePathToFullyExplicit` →`explicitfilepath.ResolvePathToFullyExplicit` (`explicitfilepath.Resolve` seemed too unclear to me)
- `private.DockerPolicyConfiguration{Identity,Namespaces}` → `policyconfiguration.DockerReferenceIdentity,Namespaces}`
- A few trivial error message, comment and commit message improvements.

OK to commit?

runcom commented:
ACK

mtrmac commented:
:+1:
